### PR TITLE
130 lineplot marginal histogram

### PR DESCRIPTION
**Pull request by moontrip**
This will address https://github.com/VEuPathDB/web-monorepo/issues/130 and https://github.com/VEuPathDB/web-monorepo/issues/131. Screenshots will be added in the following comment: tested many cases and all seem to work properly.

Two new props are added to Lineplot component, `showMarginalHistogram` and `marginalHistogramSize`. The former decides to display marginal histogram in the Lineplot Viz and the latter is for the size of marginal histogram that has default value of 0.2 (20 %): Lineplot will take 80 %. Perhaps we need to adjust `showMarginalHistogram` prop to distinguish Lineplot Viz from Lineplot with marginal histogram Viz.

Though, I am not quite sure where to place marginal histogram, top or bottom: a mockup suggested placing marginal histogram at the bottom of Lineplot, but examples we have seen and discussed in the meeting showed top area: personally, it looks better at top area.

### CODE REVIEW

**Comment by moontrip** [comment]
- ### GEMS1 with overlay using defalut binWidth
marginal histogram30


- ### same to above but unchecked 24-59 months in the legend
marginal histogram31


- ### same to above but with a facet variable
marginal histogram32


- ### Same to above but binWidth is increased and custom axis ranges are applied to check
marginal histogram4


- ### testing filter for Age variable
marginal histogram51


- ### Lineplot 0/0 case: example is taken from https://github.com/VEuPathDB/web-eda/issues/1082
marginal histogram1

**Comment by dmfalke** [comment]
> Though, I am not quite sure where to place marginal histogram, top or bottom: a mockup suggested placing marginal histogram at the bottom of Lineplot, but examples we have seen and discussed in the meeting showed top area: personally, it looks better at top area.

I think one reason to put it on the bottom is so that is it closer to the x-axis ticks. Maybe @d-callan or @asizemore can weigh in?

**Comment by d-callan** [comment]
It looks nice above. Can we duplicate the x-axis at the top of the line plot/ below the histo easily?

**Comment by moontrip** [comment]
> It looks nice above. Can we duplicate the x-axis at the top of the line plot/ below the histo easily?

Thank you for your comment @d-callan 👍 I will try to see if I can add the x-axis.

**Comment by moontrip** [comment]
@d-callan Here are some test results. Hope this will be useful for making a decision.
I have firstly checked a possibility of a tooltip described at https://github.com/VEuPathDB/web-monorepo/issues/151, but what I could achieve so far is to display a vertical line to cover up both plots. Tooltip contents are separately made depending on where a hover is located. Please see below two screenshots. Actually lineplot's tooltip contains the number of count (n).

- hovering histogram
marginal histogram xaxis2 label04

- hovering lineplot
marginal histogram xaxis2 label03

I have also tested whether we can have xaxis tick labels at the bottom of histogram. I could manage to find a way to do that, but one disadvantage is that above tooltip cannot be used with this tick labels. It is because additional xaxis tick labels require additional definition of xaxis, saying xaxis2. Anyway please see below screenshots

- placing xaxis tick labels
marginal histogram xaxis2 label01

- Above may be okay but there is a possibility that lineplot's line/marker may be overlapped with those axis tick labels in some cases: though it is most likely okay because we have set that lineplot (and scatterplot) has some internal margin at the yaxis min and max area to avoid clipping of markers around those area. 
Anyway, I found a way to avoid such a case by slightly decrease Lineplot size, like below screenshot. Total plot sizes between above and below plots are identical, but lineplot size below has smaller size than the above, as you may notice.
marginal histogram xaxis2 label02

**Comment by d-callan** [comment]
Thanks @moontrip. If we can't guarantee that both x axes are aligned then I favor the tooltip w dashed line. I think it's still ok to put the histo above. Nice job 👍

**Comment by moontrip** [comment]
> Thanks @moontrip. If we can't guarantee that both x axes are aligned then I favor the tooltip w dashed line. I think it's still ok to put the histo above. Nice job 👍

@d-callan Thank you for your comment! After testing something more, I have one update concerning the positional difference of tick labels between xaxis and xaxis2, as seen earlier. I did not test many cases yet but such a difference is probably not an issue when drawing the plot via Viz because we have set the axis range: in the storybook examples above, I did not consider the range. Here is an example at Viz with the same data used at the storybook and the tick label positions seem to be identical each other. If you prefer aligned xaxis label, then I will check more cases carefully and implement it in this PR: if this is the case, one additional question is whether you want to have a gap in between lineplot and marginal histogram like below screenshot or without the gap.

marginal histogram xaxis2 label05

**Comment by d-callan** [comment]
Can we see the regular tooltips? Remind me what they look like if we go for the duplicated x axis?

**Comment by moontrip** [comment]
> Can we see the regular tooltips? Remind me what they look like if we go for the duplicated x axis?

@d-callan Yes, we will get regular tooltips that can be seen at either Histogram Viz (hovering histogram) or Lineplot Viz (hovering lineplot) if we use two individual x-axis. Below is an example for hovering Lineplot Viz's marker.

marginal histogram xaxis2 label07

**Comment by bobular** [comment]
Reviewing now - but I suggest that linked tooltips be "future growth" i.e. another ticket.

**Comment by bobular** [comment]
Do we prefer a tiny bit of spacing between the plots?

image

**Comment by d-callan** [comment]
I think the standard tooltip w duplicate x axis is fine, and that we should start without the gap between the two plots and see how much of an issue it really is in practice.

@bobular it's already a second ticket.

File: packages/libs/components/src/plots/LinePlot.tsx [hunk 1]
```diff
@@ -270,41 +296,49 @@ const LinePlot = makePlotlyPlotComponent('LinePlot', (props: LinePlotProps) => {
...
-            // use <br> instead of \n for line break
-            return (
-              `x: ${binLabel ?? x}<br>y: ${yText}${CI}` +
-              (N !== undefined ? '<br>n: ' + N : '') +
-              (xtra ? '<br>' + xtra : '')
-            );
-          }),
-        })),
+
+        // need to exclude marginal histogram
```
Comment (bobular) [inline packages/libs/components/src/plots/LinePlot.tsx:300]: a better comment would be

```
// don't process tooltips for marginal histogram
```

But I am confused - where is `series.mode` set to non-null?  Seems like the consumer `LineplotVisualization` where the value is `'lines+markers'`.  This is a bit opaque.  Is there any way we can make it more self-explanatory?

I think at the minimum some very clear comments should be added. It seems like a specification for the props passed to this component is that if `showMarginalHistogram` is true, then a data series containing the histogram counts with undefined `mode`, must be provided.  Perhaps even validate the props to make sure this is the case.

File: packages/libs/eda/src/lib/core/components/visualizations/implementations/LineplotVisualization.tsx [hunk 2]
```diff
@@ -2348,28 +2370,107 @@ function processInputData(
...
+                  el.overlayVariableDetails.value,
+                  overlayVariable
+                )
+              : el.seriesType !== 'zeroOverZero'
+              ? 'Data'
+              : '') +
+            (el.seriesType === 'zeroOverZero'
+              ? (overlayVariable !== undefined ? ', ' : '') +
+                'Undefined Y (denominator of 0)'
+              : ''),
```
Comment (bobular) [inline packages/libs/eda/src/lib/core/components/visualizations/implementations/LineplotVisualization.tsx:2436]: Can this section be DRYed up?

**Comment by moontrip** [comment]
> I think the standard tooltip w duplicate x axis is fine, and that we should start without the gap between the two plots and see how much of an issue it really is in practice.
> 
> @bobular it's already a second ticket.

@d-callan Sounds good. I will make the change accordingly. 👍

File: packages/libs/eda/src/lib/core/components/visualizations/implementations/LineplotVisualization.tsx [hunk 3]
```diff
@@ -2348,28 +2370,107 @@ function processInputData(
...
+        el.binSampleSize &&
+        el.binSampleSize.length > 0
+      ) {
+        // compute binSampleSize one
+        const binSampleSize = categoricalMode
+          ? el.binSampleSize.map((val) =>
+              val == null
+                ? null
+                : (val as BinSampleSizeProportion).numeratorN /
+                  (val as BinSampleSizeProportion).denominatorN
```
Comment (bobular) [inline packages/libs/eda/src/lib/core/components/visualizations/implementations/LineplotVisualization.tsx:2388]: Just `denominatorN` here.  No fraction calculation.

**Review by bobular — CHANGES_REQUESTED** [review]
All looking good, but I'd like to test again and agree on some styling issues before we merge.

**Comment by bobular** [comment]
The slightly thicker top line (with no gap) bothers me in the storybook examples, but that doesn't seem to be a problem in the EDA. So I agree - no gap!

How about an axis label for the histogram?  "N"?  "Counts"?

**Comment by d-callan** [comment]
Marginal plots often don't have an axis label, but I could go either way I guess..

**Comment by moontrip** [comment]
> I have to say, it's a thing of beauty
> 
> image

@bobular Thank you for your feedbacks earlier. I will address them. By the way, do you prefer x-axis tick labels for marginal histogram or without them like your screenshot?

**Comment by d-callan** [comment]
Yes. If we don't have the axes duplicated I think the dotted line make it much more clear which values in the line plot are associated w a bin in the histo

**Review by dmfalke — COMMENTED** [review]
This is all looking great. Maybe next week we can discuss how to properly wire this up, using the `withOptions` functionality to reuse `LineplotVisualization` for the timeline plot.

**Comment by moontrip** [comment]
@dmfalke I have defined `showMarginalHistogram` at visualizationPlugins (withOptions), https://github.com/VEuPathDB/web-monorepo/pull/290/commits/c3b51c90d04888dafe92c775d95f81891c115c6d. Technically, we don't necessarily do this for general lineplot Viz because I have set the default value as `false` at `LineplotVisualization` component if `options.showMarginalHistogram` is undefined, however, it may be better to explicitly define it for clarity.

I think that we may also use this new viz (line + marginal histogram) for standalone map's Viz. Although I did define `showMarginalHistogram` in there as well, it would require some additional work as I made a note in there (`standaloneVizPlugins.ts`).

File: packages/libs/eda/src/lib/map/analysis/hooks/standaloneVizPlugins.ts [hunk 4]
```diff
@@ -40,6 +40,10 @@ export function useStandaloneVizPlugins({
       return visualization.withOptions({
         hideFacetInputs: true, // will also enable table-only mode for mosaic
         hideShowMissingnessToggle: true,
+        // TODO: need to distinguish lineplot from lineplot with marginal histogram?
+        // perhaps need to make another function only for lineplot with marginal histogram
+        // or define xxx.withOptions({}) directly at corresponding visualizationPlugins below
```
Comment (bobular) [inline packages/libs/eda/src/lib/map/analysis/hooks/standaloneVizPlugins.ts:45]: Yes I think we need to add `showMarginalHistogram` for just the appropriate lineplot(s).

**Review by bobular — CHANGES_REQUESTED** [review]
Marginal histograms are currently off everywhere.

I think we'll need a bit of back end work to define a timeline plot, right? It will need special constraints on the x axis.

**Comment by moontrip** [comment]
@bobular Thank you for your comments 👍 Yes, I disabled lineplot with marginal histogram for at pass app but perhaps we need to enable it for the time being (testing etc.). I will do that.

**Pull request merged**